File: ios/VoiceHostSoftphone/Services/Janus/JanusClient.swift
import Foundation

@MainActor
final class JanusClient: ObservableObject {
    enum ConnectionState: Equatable {
        case disconnected
        case connecting
        case connected
        case failed(String)
    }

    enum ClientError: LocalizedError {
        case notConnected
        case invalidResponse(String)
        case janusError(String)

        var errorDescription: String? {
            switch self {
            case .notConnected:
                return "Janus WebSocket is not connected."
            case .invalidResponse(let message):
                return "Invalid Janus response: \(message)"
            case .janusError(let message):
                return "Janus error: \(message)"
            }
        }
    }

    @Published private(set) var connectionState: ConnectionState = .disconnected

    var onPluginEvent: (([String: Any], [String: Any]?) -> Void)?
    var onRemoteCandidate: (([String: Any]) -> Void)?

    private let serverURL: URL
    private var socket: URLSessionWebSocketTask?
    private var receiveTask: Task<Void, Never>?
    private var keepaliveTask: Task<Void, Never>?
    private var apiSecret: String?

    private(set) var sessionID: UInt64?
    private(set) var handleID: UInt64?

    private var pending: [String: CheckedContinuation<[String: Any], Error>] = [:]

    init(serverURL: URL) {
        self.serverURL = serverURL
    }

    func connect(apiSecret: String? = nil) async throws {
        guard socket == nil else { return }
        connectionState = .connecting
        self.apiSecret = apiSecret?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty

        let configuration = URLSessionConfiguration.default
        configuration.waitsForConnectivity = true
        let session = URLSession(configuration: configuration)
        let socket = session.webSocketTask(with: serverURL, protocols: ["janus-protocol"])
        self.socket = socket
        socket.resume()

        receiveTask = Task { [weak self] in
            await self?.receiveLoop()
        }

        do {
            let create = try await request(["janus": "create"])
            guard
                let data = create["data"] as? [String: Any],
                let id = Self.uint64(data["id"])
            else {
                throw ClientError.invalidResponse("missing session id")
            }
            sessionID = id

            let attach = try await request([
                "janus": "attach",
                "plugin": AppConfig.janusPlugin,
                "session_id": id
            ])
            guard
                let attachData = attach["data"] as? [String: Any],
                let handle = Self.uint64(attachData["id"])
            else {
                throw ClientError.invalidResponse("missing SIP plugin handle id")
            }
            handleID = handle
            connectionState = .connected
            startKeepalive()
        } catch {
            connectionState = .failed(error.localizedDescription)
            disconnect()
            throw error
        }
    }

    func disconnect() {
        keepaliveTask?.cancel()
        keepaliveTask = nil
        receiveTask?.cancel()
        receiveTask = nil

        socket?.cancel(with: .goingAway, reason: nil)
        socket = nil
        sessionID = nil
        handleID = nil

        for (_, continuation) in pending {
            continuation.resume(throwing: ClientError.notConnected)
        }
        pending.removeAll()
        connectionState = .disconnected
    }

    func sendPlugin(body: [String: Any], jsep: [String: Any]? = nil) async throws {
        guard let sessionID, let handleID else { throw ClientError.notConnected }
        var message: [String: Any] = [
            "janus": "message",
            "session_id": sessionID,
            "handle_id": handleID,
            "transaction": transaction()
        ]
        message["body"] = body
        if let jsep { message["jsep"] = jsep }
        applySecret(to: &message)
        try await sendRaw(message)
    }

    func sendTrickle(candidate: [String: Any]) async throws {
        guard let sessionID, let handleID else { throw ClientError.notConnected }
        var message: [String: Any] = [
            "janus": "trickle",
            "session_id": sessionID,
            "handle_id": handleID,
            "transaction": transaction(),
            "candidate": candidate
        ]
        applySecret(to: &message)
        try await sendRaw(message)
    }

    private func request(_ base: [String: Any]) async throws -> [String: Any] {
        var message = base
        let transaction = transaction()
        message["transaction"] = transaction
        applySecret(to: &message)

        return try await withCheckedThrowingContinuation { continuation in
            pending[transaction] = continuation
            Task { [weak self] in
                guard let self else { return }
                do {
                    try await self.sendRaw(message)
                } catch {
                    if let pending = self.pending.removeValue(forKey: transaction) {
                        pending.resume(throwing: error)
                    }
                }
            }
        }
    }

    private func sendRaw(_ object: [String: Any]) async throws {
        guard let socket else { throw ClientError.notConnected }
        let data = try JSONSerialization.data(withJSONObject: object)
        guard let text = String(data: data, encoding: .utf8) else {
            throw ClientError.invalidResponse("unable to encode JSON")
        }
        try await socket.send(.string(text))
    }

    private func receiveLoop() async {
        guard let socket else { return }
        while !Task.isCancelled {
            do {
                let message = try await socket.receive()
                let text: String
                switch message {
                case .string(let value): text = value
                case .data(let data): text = String(decoding: data, as: UTF8.self)
                @unknown default: continue
                }
                handle(text)
            } catch {
                if !Task.isCancelled {
                    connectionState = .failed(error.localizedDescription)
                }
                break
            }
        }
    }

    private func handle(_ text: String) {
        guard
            let data = text.data(using: .utf8),
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }

        let janus = object["janus"] as? String

        if let transaction = object["transaction"] as? String,
           janus != "ack",
           let continuation = pending.removeValue(forKey: transaction) {
            if janus == "error" {
                let error = object["error"] as? [String: Any]
                continuation.resume(throwing: ClientError.janusError(error?["reason"] as? String ?? "unknown error"))
            } else {
                continuation.resume(returning: object)
            }
            return
        }

        if janus == "event",
           let pluginData = object["plugindata"] as? [String: Any],
           let payload = pluginData["data"] as? [String: Any] {
            onPluginEvent?(payload, object["jsep"] as? [String: Any])
            return
        }

        if janus == "trickle", let candidate = object["candidate"] as? [String: Any] {
            onRemoteCandidate?(candidate)
        }
    }

    private func startKeepalive() {
        keepaliveTask?.cancel()
        keepaliveTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(25))
                guard let self, let sessionID = self.sessionID else { return }
                var message: [String: Any] = [
                    "janus": "keepalive",
                    "session_id": sessionID,
                    "transaction": self.transaction()
                ]
                self.applySecret(to: &message)
                try? await self.sendRaw(message)
            }
        }
    }

    private func applySecret(to message: inout [String: Any]) {
        if let apiSecret { message["apisecret"] = apiSecret }
    }

    private func transaction() -> String {
        UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
    }

    private static func uint64(_ value: Any?) -> UInt64? {
        if let value = value as? UInt64 { return value }
        if let value = value as? Int { return UInt64(value) }
        if let value = value as? NSNumber { return value.uint64Value }
        return nil
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: flutter/lib/services/janus_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/io.dart';
import 'package:web_socket_channel/status.dart' as ws_status;

import '../core/app_config.dart';

typedef JsonMap = Map<String, dynamic>;

class JanusException implements Exception {
  JanusException(this.message);
  final String message;

  @override
  String toString() => message;
}

class JanusClient {
  JanusClient({required this.serverUrl});

  final Uri serverUrl;

  void Function(JsonMap payload, JsonMap? jsep)? onPluginEvent;
  void Function(JsonMap candidate)? onRemoteCandidate;
  void Function(String message)? onLog;
  void Function(Object error)? onDisconnected;

  IOWebSocketChannel? _channel;
  StreamSubscription<dynamic>? _subscription;
  Timer? _keepaliveTimer;
  final Map<String, Completer<JsonMap>> _pending = {};
  String? _apiSecret;
  int _transactionCounter = 0;

  int? sessionId;
  int? handleId;

  bool get isConnected => _channel != null && sessionId != null && handleId != null;

  Future<void> connect({String? apiSecret}) async {
    if (_channel != null) return;

    _apiSecret = apiSecret?.trim().isEmpty == true ? null : apiSecret?.trim();
    final channel = IOWebSocketChannel.connect(
      serverUrl,
      protocols: const ['janus-protocol'],
      pingInterval: const Duration(seconds: 20),
      connectTimeout: const Duration(seconds: 15),
    );
    _channel = channel;

    _subscription = channel.stream.listen(
      _handleMessage,
      onError: (Object error, StackTrace stackTrace) {
        onLog?.call('Janus WebSocket error: $error');
        _failPending(error);
        onDisconnected?.call(error);
      },
      onDone: () {
        final error = JanusException(
          'Janus WebSocket closed (${channel.closeCode ?? 'no code'})',
        );
        onLog?.call(error.message);
        _failPending(error);
        onDisconnected?.call(error);
      },
      cancelOnError: false,
    );

    try {
      await channel.ready;
      onLog?.call('Janus WebSocket connected using janus-protocol');

      final create = await _request({'janus': 'create'});
      final data = _jsonMap(create['data']);
      sessionId = _intValue(data?['id']);
      if (sessionId == null) {
        throw JanusException('Janus create response did not include a session id');
      }

      final attach = await _request({
        'janus': 'attach',
        'plugin': AppConfig.janusSipPlugin,
        'session_id': sessionId,
      });
      final attachData = _jsonMap(attach['data']);
      handleId = _intValue(attachData?['id']);
      if (handleId == null) {
        throw JanusException('Janus attach response did not include a handle id');
      }

      _startKeepalive();
      onLog?.call('Attached ${AppConfig.janusSipPlugin} handle $handleId');
    } catch (_) {
      await disconnect();
      rethrow;
    }
  }

  Future<void> disconnect() async {
    _keepaliveTimer?.cancel();
    _keepaliveTimer = null;

    final channel = _channel;
    _channel = null;
    sessionId = null;
    handleId = null;

    await _subscription?.cancel();
    _subscription = null;

    if (channel != null) {
      try {
        await channel.sink.close(ws_status.goingAway);
      } catch (_) {
        // Socket may already be closed.
      }
    }

    _failPending(JanusException('Janus client disconnected'));
  }

  Future<void> sendPlugin({
    required JsonMap body,
    JsonMap? jsep,
  }) async {
    final sid = sessionId;
    final hid = handleId;
    if (sid == null || hid == null) {
      throw JanusException('Janus SIP plugin is not connected');
    }

    final message = <String, dynamic>{
      'janus': 'message',
      'session_id': sid,
      'handle_id': hid,
      'transaction': _transaction(),
      'body': body,
    };
    if (jsep != null) message['jsep'] = jsep;
    _applySecret(message);
    _sendRaw(message);
  }

  Future<void> sendTrickle(JsonMap candidate) async {
    final sid = sessionId;
    final hid = handleId;
    if (sid == null || hid == null) {
      throw JanusException('Janus SIP plugin is not connected');
    }

    final message = <String, dynamic>{
      'janus': 'trickle',
      'session_id': sid,
      'handle_id': hid,
      'transaction': _transaction(),
      'candidate': candidate,
    };
    _applySecret(message);
    _sendRaw(message);
  }

  Future<JsonMap> _request(JsonMap base) async {
    final transaction = _transaction();
    final message = <String, dynamic>{...base, 'transaction': transaction};
    _applySecret(message);

    final completer = Completer<JsonMap>();
    _pending[transaction] = completer;
    _sendRaw(message);

    try {
      return await completer.future.timeout(const Duration(seconds: 15));
    } on TimeoutException {
      _pending.remove(transaction);
      throw JanusException('Timed out waiting for Janus response');
    }
  }

  void _sendRaw(JsonMap message) {
    final channel = _channel;
    if (channel == null) {
      throw JanusException('Janus WebSocket is not connected');
    }
    channel.sink.add(jsonEncode(message));
  }

  void _handleMessage(dynamic raw) {
    try {
      final decoded = jsonDecode(raw is String ? raw : utf8.decode(raw as List<int>));
      if (decoded is! Map) return;
      final object = Map<String, dynamic>.from(decoded);
      final janus = object['janus'] as String?;

      final transaction = object['transaction'] as String?;
      if (transaction != null && janus != 'ack') {
        final completer = _pending.remove(transaction);
        if (completer != null) {
          if (janus == 'error') {
            final error = _jsonMap(object['error']);
            completer.completeError(
              JanusException(error?['reason']?.toString() ?? 'Janus error'),
            );
          } else {
            completer.complete(object);
          }
          return;
        }
      }

      if (janus == 'event') {
        final pluginData = _jsonMap(object['plugindata']);
        final payload = _jsonMap(pluginData?['data']);
        if (payload != null) {
          onPluginEvent?.call(payload, _jsonMap(object['jsep']));
        }
        return;
      }

      if (janus == 'trickle') {
        final candidate = _jsonMap(object['candidate']);
        if (candidate != null) onRemoteCandidate?.call(candidate);
        return;
      }

      if (janus == 'webrtcup' || janus == 'hangup' || janus == 'media') {
        onLog?.call('Janus event: $object');
      }
    } catch (error) {
      onLog?.call('Unable to parse Janus message: $error');
    }
  }

  void _startKeepalive() {
    _keepaliveTimer?.cancel();
    _keepaliveTimer = Timer.periodic(const Duration(seconds: 25), (_) {
      final sid = sessionId;
      if (sid == null || _channel == null) return;
      final message = <String, dynamic>{
        'janus': 'keepalive',
        'session_id': sid,
        'transaction': _transaction(),
      };
      _applySecret(message);
      try {
        _sendRaw(message);
      } catch (_) {
        // The socket listener will report the connection failure.
      }
    });
  }

  void _applySecret(JsonMap message) {
    final secret = _apiSecret;
    if (secret != null && secret.isNotEmpty) message['apisecret'] = secret;
  }

  String _transaction() {
    _transactionCounter++;
    return '${DateTime.now().microsecondsSinceEpoch.toRadixString(36)}${_transactionCounter.toRadixString(36)}';
  }

  void _failPending(Object error) {
    final pending = List<Completer<JsonMap>>.from(_pending.values);
    _pending.clear();
    for (final completer in pending) {
      if (!completer.isCompleted) completer.completeError(error);
    }
  }

  static JsonMap? _jsonMap(dynamic value) {
    if (value is Map) return Map<String, dynamic>.from(value);
    return null;
  }

  static int? _intValue(dynamic value) {
    if (value is int) return value;
    if (value is num) return value.toInt();
    return int.tryParse(value?.toString() ?? '');
  }
}
